Add Bitcoin blockchain analysis tutorial (#1176)
timescale/docs: timescaledb/tutorials/analyze-bitcoin-blockchain/analyze-blockchain.md

queries faster by continuously materializing aggregated data. At the same
time, they provide real-time results. That means they include the latest data from the
underlying hypertable.

By using continuous aggregates, you simplify and speed up your queries.

In this tutorial, you learn to create three continuous aggregates, focusing
on three aspects of the dataset:
* Bitcoin transactions
* Bitcoin blocks
* Coinbase transactions (miner revenue)

Pre-aggregating your data is important because the dataset contains a lot
of transactions: over 10,000 per hour.

In this section, you also learn how to keep your continuous
aggregate views up-to-date with automatic refresh policies.

## Hyperfunctions for simplified statistical queries
In the following queries you can find custom SQL functions that are not part of
vanilla PostgreSQL. These queries are TimescaleDB hyperfunctions and they are
either part of the TimescaleDB extension or the Toolkit extension. You need to
install and enable the Toolkit extension to successfully run the following
queries.

Enable Toolkit extension:
```sql
CREATE EXTENSION timescaledb_toolkit;
```

> YOu have everything now to create the continuous aggregates and start analyzing!

Reviewer: ```suggestion
You have everything now to create the continuous aggregates and start analyzing!
```

* `min_miner_revenue`: Minimum miner revenue received after mining a block per day
* `max_miner_revenue`: Maximum miner revenue received after mining a block per day

Add a refresh policy to keep the continuous aggregate up-to-date:
```sql
SELECT add_continuous_aggregate_policy('one_hour_coinbase',
   start_offset => INTERVAL '3 hours',
   end_offset => INTERVAL '1 hour',
   schedule_interval => INTERVAL '1 hour');
```

In each continuous aggregate definition, the `time_bucket()` function controls
how large the time buckets are. The examples all use 1-hour time
buckets.

## Generate insights with SQL
Here are some questions you might ask about blockchain
transactions, blocks, and miner revenue. For each question,
you get a relevant SQL query and a chart that answers the question.

**Questions**
* Is there any connection between the number of transactions and the transaction fees?
* Does the transaction volume affect the BTC-USD rate?
* Do more transactions in a block mean the block is more expensive to mine?
* What percentage of the average miner's revenue comes from fees vs. block rewards?
* How does block weight affect miner fees?
* What’s the average miner revenue per block?

> 

Reviewer: This feels unnecessary to me since we don't link to anything. Any other opinions?

Author: I can add anchor links so the reader can jump to the one they find the most interesting

Author: https://github.com/timescale/docs/pull/1176/commits/250b638967d8539d5237a70b48dd1a66c186ff20

### How does block weight affect miner fees?
You’ve already seen that more transactions in a block mean it’s more expensive
to mine. Is it the same with block weights? The more transactions a block has,
the larger its size (or weight), so the block weight and mining fee should
be tightly correlated. 

This query uses a 12-hour moving average to calculate the block weight and
block mining fee over time.

```sql
WITH stats AS (
   SELECT
       bucket,
       stats_agg(block_weight, block_fee_sat) AS block_stats
   FROM one_hour_blocks
   WHERE bucket > NOW() - INTERVAL '5 days'
   GROUP BY bucket
)
SELECT
   bucket as "time",
   average_y(rolling(block_stats) OVER (ORDER BY bucket RANGE '12 hours' PRECEDING)) AS "block weight",
   average_x(rolling(block_stats) OVER (ORDER BY bucket RANGE '12 hours' PRECEDING))*0.00000001 AS "mining fee"
FROM stats
ORDER BY 1
```

![block weight and fees](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/weight_fees.png)

You can see that the block weight and block mining fee are indeed tightly
connected to each other. In practice, you can also see that four million

> weight units are the size limit introduced in the 2017’s SegWit update.

Reviewer 2: ```suggestion
weight units is the size limit introduced in the 2017 SegWit update.
```

This means that, looking at this graph, there’s still room to grow for
individual blocks, and they could include even more transactions.

### What’s the average miner revenue per block?
Now, analyze how much revenue miners actually generate by mining a
new block on the blockchain, including fees and block rewards. This query
analyzes the last day with 12-hour moving averages.

```sql
SELECT
   bucket as "time",
   average_y(rolling(stats_miner_revenue) OVER (ORDER BY bucket RANGE '12 hours' PRECEDING))*0.00000001 AS "revenue in BTC"
FROM one_hour_coinbase
WHERE bucket > NOW() - INTERVAL '1 day'
ORDER BY 1
```
![Average miner revenue per block, plotted over the last day](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/miner_revenue_per_block.png)

> (THIS CHART NEEDS AN UPDATE - larger time frame?)

Reviewer 2: don't know if you're ready to merge but if you are just a friendly reminder that this is here :)

Author: Thx I missed that, the chart is fine just needed to fix the SQL code

more than ten
thousand Bitcoin transactions in an hour so it's essential to aggregate
them into larger time buckets (1-hour, 12-hours, 1-day, etc) for analysis.

This tutorial creates three continuous aggregates, focusing on three aspects
of the dataset:
* Bitcoin transactions
* Bitcoin blocks
* Coinbase transactions (miner revenue)

By using continuous aggregates, your analytical and
time-series queries are vastly simplified and
sped up going forward.

This tutorial, in a later section, also shows you how to keep these continuous
aggregate views up-to-date automatically with refresh policies.

## Hyperfunctions for simplified statistical queries
In the following queries you can find custom SQL functions that are not part of
vanilla PostgreSQL. These queries are TimescaleDB hyperfunctions and they are
either part of the TimescaleDB extension or the Toolkit extension. You need to
install and enable the Toolkit extension to successfully run the following
queries.

Enable Toolkit extension:
```sql
CREATE EXTENSION timescaledb_toolkit;
```

YOu have everything now to create the continuous aggregates and start analyzing!

> 

Author: agree!

> 

Author: https://github.com/timescale/docs/pull/1176/commits/4faf19c6e7ac5f928fec8d87bfe4371b127f23f8

On this chart, the green line indicates the average transaction volume over
time, and the yellow line indicates the average fee per transaction over
time. This type of insight can help you, for example, decide whether you
should submit a transaction now or wait a couple of days for fees to
decrease. Now let’s look at an analysis that traders might find insightful.

### Does the transaction volume affect the BTC-USD rate?
In cryptocurrency trading, there’s a lot of speculation. Finding correlations
between fundamental blockchain metrics, like transaction volume or fees and the
BTC-USD rate can help traders shape their strategy and make decisions based
on data.

```sql
SELECT
 bucket AS "time",
 tx_count as "tx volume",
 total_fee_usd / (total_fee_sat*0.00000001) AS "btc-usd rate"
FROM one_hour_transactions
WHERE bucket > NOW() - INTERVAL '1 day'
ORDER BY 1
```
![volume and BTC-USD](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/volume_btc_usd.png)

Again, the green line shows the average transaction volume over time and the
yellow line shows the BTC-USD conversion rate at any given time.

> Next, let's segue into block-level insights by analyzing the connection

Reviewer 2: ```suggestion
Next, get block-level insights by analyzing the connection
```

between transactions and blocks.

### Do more transactions in a block mean the block is more expensive to mine?
For someone who is not an expert on how the Bitcoin blockchain works, it might be a

> difficult question to answer. So let’s ask the data and see if it’s true.

Reviewer 2: Why is this necessary? Is this something that is often claimed, but that is not intuitive to see? If so, I would just say that. Something like:

`It's often said that a block with more transactions is more expensive to mine. Analyze the data to see if it supports this trend.`

> For this kind of analysis you might want to look at a larger time frame,
> so let’s change the analyzed time range to the last five days.

Reviewer 2: ```suggestion
For this kind of analysis, you might want to look at a larger time frame.
Change the analyzed time range to the last five days.
```

```sql
SELECT
 bucket as "time",
 avg(tx_count) AS transactions,
 avg(block_fee_sat)*0.00000001 AS "mining fee"
FROM one_hour_blocks
WHERE bucket > now() - INTERVAL '5 day'
GROUP BY bucket
ORDER BY 1
```

> ![transactions in block and mining fee](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/tx_in_block_expensive.png)

Reviewer 2: ```suggestion
![Line graph with two lines showing the average number of transactions in a block and the block mining fee, over the last five days](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/tx_in_block_expensive.png)
```

Unsurprisingly, there’s a high correlation between the number of transactions
in a block and the mining fee. The more transactions a block has, the higher
the block mining fee.

> Let’s dive deeper and see if there is the same correlation between block

Reviewer 2: ```suggestion
In the next query, see if there is the same correlation between block
```

weight and mining fee. (Block weight is the size measure of a block). More
transactions should increase the block weight, boosting the miner fee as well.
This query looks very similar to the previous one:
```sql
SELECT
 bucket as "time",
 avg(block_weight) as "block weight",
 avg(block_fee_sat*0.00000001) as "mining fee"
FROM one_hour_blocks
WHERE bucket > now() - INTERVAL '5 day'
group by bucket
ORDER BY 1
```

> ![block weight and mining fee](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/weight_fee.png)

Reviewer 2: ```suggestion
![Line graph with two lines showing the block weight and the block mining fee, over the last five days](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/weight_fee.png)
```
Same comment as for last graph

> And yes—you can see the same kind of high correlation between block weight
> (defined in weight units) and mining fee. Except when the block weight gets
> close to its maximum value (4M weight units) in which case it's impossible for a
> block to include more transactions.

Reviewer 2: ```suggestion
You can see the same kind of high correlation between block weight
(defined in weight units) and mining fee. The relationship weakens when the block weight gets
close to its maximum value (4 million weight units), in which case it's impossible for a
block to include more transactions.
```

> In the previous charts, you saw how mining fees are correlated to block
> weights, transaction volumes, etc. But now, let’s analyze the data from a
> different perspective. Miner revenue is not only made up of miner fees, it
> also includes block rewards after mining a new block. This reward is currently
> 6.25 BTC and it gets halved every four years. Let’s analyze miner revenue!

Reviewer 2: ```suggestion
In the previous charts, you saw how mining fees are correlated to block
weights and transaction volumes. In the next query, analyze the data from a
different perspective. Miner revenue is not only made up of miner fees. It
also includes block rewards after mining a new block. This reward is currently
6.25 BTC, and it gets halved every four years. What are some trends in miner revenue?
```

> ### What percentage of the average miner's revenue comes from fees vs. block rewards?

Reviewer 2: ```suggestion
### What percentage of the average miner's revenue comes from fees compared to block rewards?
```

Miners are incentivized to keep the network up and running because they earn

> fees and rewards after mining each block. But let’s look at how much of their
> revenue comes from these two sources.

Reviewer 2: ```suggestion
fees and rewards after mining each block. How much of their
revenue comes from each source?
```

```sql
WITH coinbase AS (
   SELECT block_id, output_total AS coinbase_tx FROM transactions
   WHERE is_coinbase IS TRUE and time > NOW() - INTERVAL '5 days'
)
SELECT
   bucket as "time",
   avg(block_fee_sat)*0.00000001 AS "fees",
   FIRST((c.coinbase_tx - block_fee_sat), bucket)*0.00000001 AS "reward"
FROM one_hour_blocks b
INNER JOIN coinbase c ON c.block_id = b.block_id
GROUP BY bucket
ORDER BY 1;
```

> ![fees and block reward](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/revenue_ratio.png)

Reviewer 2: ```suggestion
![Line graph with two lines showing the average fee and block reward, over the last five days](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/revenue_ratio.png)
```

> (THIS CHART NEEDS AN UPDATE)

Author: ```suggestion
```

> This chart analyzes the last 5 days of average miner revenue. The left
> axis is a percentage value that indicates the portion of total revenue that
> comes from transaction fees (green) and block rewards (yellow). It might be
> surprising for some, but most miner revenue comes from block rewards
> (6.25BTC at the moment). The fees portion never went above 3 % in the
> last 5 days. 

Reviewer 2: ```suggestion
This chart analyzes the last five days of average miner revenue. The left
axis shows the percentage of total revenue that
comes from transaction fees (green) and block rewards (yellow). Most miner
revenue actually comes from block rewards
(6.25&nbsp;BTC at the moment). Fees never accounted for more than 3% in the
last five days. 
```

Five or 5, doesn't really matter, just make it consistent. I've seen five everywhere else so went with that

This kind of analysis can start discussions around the long-term fading of

> block rewards and how on-chain fees will need to go up in the future to
> incentivize the miners and sustain the network. (Note that the left axis is

Reviewer 2: ```suggestion
block rewards and how on-chain fees need to rise to
incentivize miners and sustain the network. (Note that the left axis is
```

logarithmic-scale, so it’s easier to see the green "fees" portion.)

### How does block weight affect miner fees?
You’ve already seen that more transactions in a block mean it’s more expensive
to mine. Is it the same with block weights? The more transactions a block has,
the larger its size (or weight), so the block weight and mining fee should
be tightly correlated. 

This query uses a 12-hour moving average to calculate the block weight and
block mining fee over time.

```sql
WITH stats AS (
   SELECT
       bucket,
       stats_agg(block_weight, block_fee_sat) AS block_stats
   FROM one_hour_blocks
   WHERE bucket > NOW() - INTERVAL '5 days'
   GROUP BY bucket
)
SELECT
   bucket as "time",
   average_y(rolling(block_stats) OVER (ORDER BY bucket RANGE '12 hours' PRECEDING)) AS "block weight",
   average_x(rolling(block_stats) OVER (ORDER BY bucket RANGE '12 hours' PRECEDING))*0.00000001 AS "mining fee"
FROM stats
ORDER BY 1
```

> ![block weight and fees](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/weight_fees.png)

Reviewer 2: ```suggestion
![block weight and fees](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/weight_fees.png)
```
Am I misinterpreting something, or is this not the graph produced from the query? Graph is titled 4-week moving average, and x-axis spans less than a day

Also this section seems very similar to the weight vs. fee section above, but using a rolling average. Is there a reason for having both, or should they be consolidated into one?

Author: correct good catch let me create the proper chart

You can see that the block weight and block mining fee are indeed tightly
connected to each other. In practice, you can also see that four million
weight units are the size limit introduced in the 2017’s SegWit update.
This means that, looking at this graph, there’s still room to grow for
individual blocks, and they could include even more transactions.

### What’s the average miner revenue per block?

> Now let’s analyze how much revenue miners actually generate by mining a
> new block on the blockchain, including fees and block rewards. This query

Reviewer 2: ```suggestion
Now, analyze how much revenue miners actually generate by mining a
new block on the blockchain, including fees and block rewards. This query
```

analyzes the last day with 12-hour moving averages.

```sql
SELECT
   bucket as "time",
   average_y(rolling(stats_miner_revenue) OVER (ORDER BY bucket RANGE '12 hours' PRECEDING))*0.00000001 AS "revenue in BTC"
FROM one_hour_coinbase
WHERE bucket > NOW() - INTERVAL '1 day'
ORDER BY 1
```

> ![miner revenue per block](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/miner_revenue_per_block.png)

Reviewer 2: ```suggestion
![Average miner revenue per block, plotted over the last day](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/miner_revenue_per_block.png)
```

(THIS CHART NEEDS AN UPDATE - larger time frame?)

> To add some spice to the chart, let’s add the BTC-USD rate to the analysis

Reviewer 2: ```suggestion
To make the chart more interesting, add the BTC-USD rate to the analysis
```

and increase the time range:

```sql
SELECT
   bucket as "time",
   average_y(rolling(stats_miner_revenue) OVER (ORDER BY bucket RANGE '7 days' PRECEDING))*0.00000001 AS "revenue in BTC",
   average_x(rolling(stats_miner_revenue) OVER (ORDER BY bucket RANGE '7 days' PRECEDING)) AS "revenue in USD"
FROM coinbase_one_day
WHERE
 bucket > NOW() - INTERVAL '5 day'
ORDER BY 1
```

> ![miner revenue per block](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/miner_revenue_per_block_with_btcusd.png)

Reviewer 2: ```suggestion
![Average miner revenue per block, plotted in BTC and USD, over the last five days](https://assets.timescale.com/docs/images/tutorials/bitcoin-blockchain/miner_revenue_per_block_with_btcusd.png)
```
